lib/tic_tac_toe/board_protocols.ex

```diff
@@ -0,0 +1,15 @@
+defprotocol TicTacToe.BoardUpdate do
+  def place_token(board, position, token)
+end
+
+defprotocol TicTacToe.BoardState do
+  def available_positions(board)
+end
+
+defprotocol TicTacToe.BoardInspect do
+  def get(board, position)
+end
+
+defprotocol TicTacToe.BoardOutcome do
+  def game_over?(board)
+end
```

Comment:
I like that you explored ISP here. I am curious if there are any new usage patterns that you see. One pattern that I noticed is that `BoardUpdate` and `BoardOutcome` are only used once and they are used together. This makes me wonder a) if they should be a part of the same protocol or b) they should be combined into a single function.

Secondly, `BoardState` and `BoardInspect` feel very similar. Do they need to be separate protocols? I am also wondering if maybe we could get by with just `BoardInspect`? I *think* the only usage of `BoardState` could be achieved with just `BoardInspect`.

One thing to consider with ISP is whether or not we are willing to introduce a new abstraction. Premature abstraction can be much much harder to back out of later.

Reply:
I thought of bundling some of these protocols, but we face the same problems with unused functions.  We only use the get function in the Presenter, and we use available positions in the Player.  If we bundled them we would be in the same quagmire of including functions we don't need.  Likewise, with outcome, we will be using this one in the presenter, but if we bundle place_token, then again we have a function we don't use.